utils: managed_bytes: optimize memory usage for small buffers

michoecho opened the pull request:
managed_bytes is implemented as chain of blob_storage objects.
Each blob_storage contains 24 bytes of metadata. But in the most
common case -- when there is only a single element in the chain --
16 bytes of this metadata is trivial/unused.

This is regrettable waste because managed_bytes is used for every
database cell in the memtables and cache. It means that every value
of size >= 7 bytes (smaller ones fit in the inline storage of
managed_bytes) receives 16 bytes of useless overhead.

To correct that, this series adds to managed_bytes an alternative storage
layout -- used for buffers small enough to fit in one fragment -- which only
stores the necessary minimum of metadata. (That is: a pointer to the parent,
to facilitate moving the storage during memory defragmentation).

This saves 16 bytes on every cell greater than 15 bytes. Which includes e.g.
every live cell with value bigger than 6 bytes, which likely applies to most cells.

Before:
```
$ build/release/scylla perf-simple-query --duration 10
median 218692.88 tps ( 61.1 allocs/op,  13.1 tasks/op,   41762 insns/op,        0 errors)
$ build/release/scylla perf-simple-query --duration 10 --write
median 173511.46 tps ( 58.3 allocs/op,  13.2 tasks/op,   53258 insns/op,        0 errors)
$ build/release/test/perf/mutation_footprint_test -c1 --row-count=20 --partition-count=100 --data-size=8 --column-count=16
 - in cache:     2580222
 - in memtable:  2549852
```

After:
```
$ build/release/scylla perf-simple-query --duration 10
median 218780.89 tps ( 61.1 allocs/op,  13.1 tasks/op,   41763 insns/op,        0 errors)
$ build/release/scylla perf-simple-query --duration 10 --write
median 173105.78 tps ( 58.3 allocs/op,  13.2 tasks/op,   52913 insns/op,        0 errors)
$ build/release/test/perf/mutation_footprint_test -c1 --row-count=20 --partition-count=100 --data-size=8 --column-count=16
 - in cache:     2068238
 - in memtable:  2037696
```

mykaul commented on utils/managed_bytes.hh:
```diff
@@ -67,32 +87,137 @@ struct blob_storage {
+//    Used for data which fits into max_inline_size.
+// 2. External contiguous. (Single-allocation).
+//    Used for data which fits into preferred_max_contiguous_allocation().
+//    (At the moment of writing: 128 kiB and 12,8 kiB in LSA).
```
Not sure what '12,8 KiB' mean here?

dorlaor commented:
Wow

On Thu, Jun 15, 2023 at 5:41 PM Michał Chojnowski ***@***.***>
wrote:

> utils: managed_bytes: optimize memory usage for small buffers
>
> managed_bytes is implemented as chain of blob_storage objects.
> Each blob_storage contains 24 bytes of metadata. But in the most
> common case -- when there is only a single element in the chain --
> 16 bytes of this metadata is trivial/unused.
>
> This is regrettable waste because managed_bytes is used for every
> database cell in the memtables and cache. It means that every value
> of size >= 7 bytes (smaller ones fit in the inline storage of
> managed_bytes) receives 16 bytes of useless overhead.
>
> To correct that, this series adds to managed_bytes an alternative storage
> layout -- used for buffers small enough to fit in one fragment -- which
> only
> stores the necessary minimum of metadata. (That is: a pointer to the
> parent,
> to facilitate moving the storage during memory defragmentation).
>
> This saves 16 bytes on every cell greater than 15 bytes. Which includes
> e.g.
> every live cell with value bigger than 6 bytes, which likely applies to
> most cells.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/scylladb/scylladb/pull/14263
> Commit Summary
>
>    - 70a9514
>    <https://github.com/scylladb/scylladb/pull/14263/commits/70a9514384267936324f2527819e928fac7fdbc6>
>    utils: managed_bytes: rewrite managed_bytes methods in terms of
>    managed_bytes_view
>    - d2cf4d5
>    <https://github.com/scylladb/scylladb/pull/14263/commits/d2cf4d5c40958c5bf15f6109df59e01900dae991>
>    utils: managed_bytes: optimize memory usage for small buffers
>    - ee96e47
>    <https://github.com/scylladb/scylladb/pull/14263/commits/ee96e47bd73d614a771972adf71dbf170be333c3>
>    utils: managed_bytes: optimize memory usage for small buffers
>
> File Changes
>
> (3 files <https://github.com/scylladb/scylladb/pull/14263/files>)
>
>    - *M* utils/managed_bytes.cc
>    <https://github.com/scylladb/scylladb/pull/14263/files#diff-7f2b4533cad6444c6934cc58dcb997024943b9b18e7f420604ad539d8771d001>
>    (12)
>    - *M* utils/managed_bytes.hh
>    <https://github.com/scylladb/scylladb/pull/14263/files#diff-5d4d64bcc5a9e030a442609b7639e8c33475e722aa2849a01360b7a91081008d>
>    (387)
>    - *M* utils/mutable_view.hh
>    <https://github.com/scylladb/scylladb/pull/14263/files#diff-313a98675cd5173f099487878d494297ea90e16000ebf994362b05937ea3945d>
>    (3)
>
> Patch Links:
>
>    - https://github.com/scylladb/scylladb/pull/14263.patch
>    - https://github.com/scylladb/scylladb/pull/14263.diff
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/scylladb/scylladb/pull/14263>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANHUROMLKOCFOVVA7A4NFLXLMNKFANCNFSM6AAAAAAZH6Y3JU>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1860/

denesb commented on utils/managed_bytes.hh:
```diff
@@ -72,27 +72,76 @@ struct blob_storage {
     struct small_blob {
         bytes_view::value_type data[max_inline_size];
         int8_t size; // -1 -> use blob_storage
+                     // -2 -> use medium_blob_storage
```
I thinks `size` should be hoisted out of the `small_blob`, and then we wouldn't need to play games with type punning.
Also a nit: I find it counter-intuitive that we have medium, big and tiny (inline), in other words I would expect big to get the size of -2 and medium the size of -1.
Also we could take this opportunity to use better naming: inline, single-chunk and multi-chunk, instead of opaque big and medium.

michoecho replied:
> Also a nit: I find it counter-intuitive that we have medium, big and tiny (inline), in other words I would expect big to get the size of -2 and medium the size of -1.

The next (cleanup) patch swaps them, and the end result is as you say.

> Also we could take this opportunity to use better naming: inline, single-chunk and multi-chunk, instead of opaque big and medium.

Alright, will do.

> I thinks size should be hoisted out of the small_blob, and then we wouldn't need to play games with type punning.

The next patch puts it at the beginning of all 3 structs, and at that point it's legal because of "common initial sequence" rules.
It could be hoisted out, but that doesn't really change anything at all.

denesb replied:
> It could be hoisted out, but that doesn't really change anything at all.

It does change things: we no longer need to have these tricky `padding` fields, the usage of which seems very fragile to me. The simpler, the better.

denesb replied:
We may not even need wrapper structs any more (maybe for documentation purposes only).

michoecho replied:
> It does change things: we no longer need to have these tricky padding fields, the usage of which seems very fragile to me.

The `padding` is just padding. They aren't needed for correctness. I added a comment there explaining that it's only there to make the alignment more natural, in a (very faint) hope that it helps the compiler slightly. (E.g. it puts both `medium` and `big` pointers in the same place, so the instruction which reads it can be common for both branches).

By the way, we can put the padding bytes to work. I have an idea that we could use it to store an allocator signature (e.g. a LSA vs nonLSA bit, or a LSA region ID), for hardening against alloc/free mismatch bugs (https://github.com/scylladb/scylladb/issues/13216).

> We may not even need wrapper structs any more (maybe for documentation purposes only).

I think having a wrapper struct for each variant is nice in any usage of union. I don't see merit in avoiding the wrappers.

denesb replied:
> The padding fields are just padding. They aren't needed for correctness.

Well, if you add too much of it, the size of the union will be bloated.

denesb replied:
If you want to use the padding fields, just add regular fields instead.

michoecho replied:
That's why there is exactly enough padding to make `big` and `medium` as big as `small`.

michoecho replied:
> If you want to use the padding fields, just add regular fields instead.

Yes. My point was: you need a wrapper struct for that.

denesb replied:
I don't mind the wrapper structs. The changes I requested are:
* Hoist size into the `managed_bytes` class itself.
* Remove the padding fields, which have no role anymore.

michoecho replied:
> I don't mind the wrapper structs. The changes I requested are:
> 
>     * Hoist size into the `managed_bytes` class itself.
> 
>     * Remove the padding fields, which have no role anymore.

Alright.

avikivity replied:
It may be difficult to separate size out. We end up with a 15-byte structure, which the compilers don't like to generate. Maybe [[gnu::packed]] helps there.

michoecho replied:
> It may be difficult to separate size out. We end up with a 15-byte structure, which the compilers don't like to generate. Maybe [[gnu::packed]] helps there.

Yes, struct packing reduces its alignment to 1. (It would be weird if it didn't).

mykaul replied:
Any perf implications due to packing?

michoecho replied:
> Any perf implications due to packing?

I'm not sure what you are asking about. Unaligned accesses? These are supposed to have no penalty on modern CPUs.

mykaul replied:
Yes, unaligned. It has no penalty as long as it's all in the same cacheline, no?

michoecho replied:
> Yes, unaligned. It has no penalty as long as it's all in the same cacheline, no?

Even if they are in different cachelines.

https://lemire.me/blog/2012/05/31/data-alignment-for-speed-myth-or-reality/

There is some penalty when it crosses page boundaries, though. 

Anyway, I don't think any of this matters. Scylla lives in a different order of magnitude.

avikivity commented on utils/managed_bytes.hh:
```diff
@@ -104,41 +139,82 @@ struct [[gnu::packed]] medium_blob_storage {
-        medium_blob_storage::ref_type ref;
-        int8_t padding[4];
+        int8_t inline_size; // In layout 2, always equals -1.
+        int8_t padding[3]; // Only used to pad `ref` to its natural alignment. (Could help with codegen).
```
7 would be better, perhaps another patch.

michoecho replied:
No, because the medium_blob_storage::ref_type is 12 bytes, not 8 bytes (it contains not just the pointer, but also the size.

It can't be hoisted outside because the medium_blob_storage has to be able to find out its size during LSA compaction.

avikivity commented on utils/managed_bytes.hh:
```diff
@@ -104,41 +139,82 @@ struct [[gnu::packed]] medium_blob_storage {
-        blob_storage::ref_type ptr;
-        int8_t padding[8];
+        int8_t inline_size; // In layout 3, always equals -2.
+        int8_t padding[7]; // Unused.
```
7!

avikivity commented:
> It seems that reimplementing the specialized `managed_bytes(const managed_bytes& o)` with the general view-based constructor causes a noticeable regression in `perf_simple_query`. (~43800 instr/op -> ~45200 instr/op). I'll have to take a closer look at why that's the case.

Wow.

michoecho commented:
> Looks good.
> 
> We added some interoperability between bytes_ostream and managed_bytes, does it need adjustment?

No, bytes_ostream just unconditionally uses the multi-chunk layout (which is the same as before, so bytes_ostream logic doesn't have to change), even if it there is only a single chunk.
IIUC the output of bytes_ostream is never stored long-term in LSA, so the waste of several bytes there isn't worth improving.

michoecho commented:
v2:
- fixed typos in comments
- hoisted `inline_size` outside the union
- removed padding fields in the union variants
- ditched wrapper structs over union variants
- brought back fast paths (for the small variants) for the copy constructor and operator[] of managed_bytes, because they matter in microbenchmarks
- renamed "small, medium, big" to "inline, single-chunk, multi-chunk" everywhere

> > It seems that reimplementing the specialized `managed_bytes(const managed_bytes& o)` with the general view-based constructor causes a noticeable regression in `perf_simple_query`. (~43800 instr/op -> ~45200 instr/op). I'll have to take a closer look at why that's the case.
> 
> Wow.

I didn't dig deep, but from what I gathered, the difference had mostly to do with copying clustering keys, which in `perf_simple_query` fit into the inline storage, so they benefited from the fast path which only needs like 4 instructions. (I checked that by growing the keys to 16 bytes — it made the difference go away.)
The difference seems excessive (is the general path so expensive, or are the keys copied so much?) but I didn't want to spend too much time on this — it's not that interesting — so I just brought back the fast path.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1988/

michoecho commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1988/

I got defeated by a C++ gotcha here — I forgot that when delegating constructors are used, the object is considered constructed as soon as the delegating constructor finishes. But I used a delegating constructor in the copy constructor I just added, and when the body of the copy constructor threw (in a bad_alloc test), the destructor was called unexpectedly (for me).

denesb reviewed (approved):
Nice.

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1992/

michoecho commented:
> CI state FAILURE - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1992/

And here I was just sloppy. The newest force-push should be correct.

avikivity commented on utils/managed_bytes.hh:
```diff
@@ -68,34 +88,117 @@ struct blob_storage {
+        bytes_view::value_type inline_data[max_inline_size]; // Stores the data directly. Size is in inline_size.
+        single_chunk_blob_storage::ref_type single_chunk_ref; // Points to external storage and stores the data size.
+        multi_chunk_blob_storage::ref_type multi_chunk_ref; // Points to external storage.
     } _u;
```
I clang able to optimize moves to a couple of instructions (merging the copying of _inline_size and _u)? Please verify manually.

Please add perf-simple-query report to the cover letter, both read and write, this is critical infrastructure.

michoecho replied:
Alright.

avikivity replied:
Still missing. Also, please rebase to destale the test results.

michoecho replied:
> Still missing. Also, please rebase to destale the test results.

Rebased and added the test results to the cover letter.

> I clang able to optimize moves to a couple of instructions (merging the copying of _inline_size and _u)? Please verify manually.

No, in fact the compiler insist on copying `_inline_size` and `_u` separately, even though their loads and stores are begging to be merged. I don't know why.

I put a manual memcpy in the move constructor to fix that. (But does it really matter?)

avikivity commented:
Please also add memory_footprint_test report to the cover letter.

No, I don't mind if the cover letter is much larger than the patch.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1996/

mykaul commented:
@michoecho - can you push this forward, so we'll get it into 5.4?

michoecho commented:
> @michoecho - can you push this forward, so we'll get it into 5.4?

I could, but do we really want it? The improvement is minor. Maybe we shouldn't do any minor improvements until we are comfortable with stability?

mykaul commented:
> > @michoecho - can you push this forward, so we'll get it into 5.4?
> 
> I could, but do we really want it? The improvement is minor. Maybe we shouldn't do any minor improvements until we are comfortable with stability?

It can certainly wait post 5.4 branching then.

avikivity commented:
> > @michoecho - can you push this forward, so we'll get it into 5.4?
> 
> I could, but do we really want it? The improvement is minor. Maybe we shouldn't do any minor improvements until we are comfortable with stability?

I don't see a reason to block it, it will just get stale.

bhalevy commented:
ping, can we merge this change?

michoecho commented:
> ping, can we merge this change?

We could, it's just that the change is lackluster. We can do better.

Every cell bigger than 15 bytes (which includes every live cell where the value is bigger than 6 bytes) has 40 bytes of overhead. This PR lowers the overhead to 24 bytes. But what we really want is ~2 bytes, which you can get with a data structure which groups multiple cells within one allocation. It shouldn't be very hard to code, we just never got to it.

Note: I just noticed that the patch is missing a relevant scylla-gdb.py update.

mykaul commented:
> We could, it's just that the change is lackluster. We can do better.

Don't let Perfect be the enemy of good - let's get it in!

denesb commented:
@michoecho please rebase, address https://github.com/scylladb/scylladb/pull/14263#issuecomment-1602347945 and this PR is good to go. You can always improve on it in a follow-up.

mykaul commented:
Reminds me of this comment:
```
class data_value {
    void* _value;  // FIXME: use "small value optimization" for small types
    data_type _type;
```

(from types/types.hh )

mykaul commented:
@avikivity - can you complete the review so we can get it merged in time for 6.0?

denesb commented:
> Reminds me of this comment:
> 
> ```
> class data_value {
>     void* _value;  // FIXME: use "small value optimization" for small types
>     data_type _type;
> ```
> 
> (from types/types.hh )

This is not an important optimization to do. `data_value` objects are only used transiently when either building mutations or when building queries. They are not used in any code where we store data in memory longer term.

avikivity commented:
> > Reminds me of this comment:
> > ```
> > class data_value {
> >     void* _value;  // FIXME: use "small value optimization" for small types
> >     data_type _type;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > (from types/types.hh )
> 
> This is not an important optimization to do. `data_value` objects are only used transiently when either building mutations or when building queries. They are not used in any code where we store data in memory longer term.

IIRC they're not used in the data path, except perhaps when evaluating functions.

avikivity commented on bytes_ostream.hh:
```diff
@@ -37,11 +37,11 @@ private:
     //       and chunk::frag_size refers to the currently occupied space in the chunk.
     //       After building, the first chunk::size is the whole object size, and chunk::frag_size
     //       doesn't change. This fits with managed_bytes interpretation.
```
> of size >= 7 bytes (smaller ones fit in the inline storage of

16? max_inline_size = 15.

michoecho replied:
And that's assuming you don't have TTL, because if you do, then you get another 12 bytes.

avikivity replied:
Ah, such a pity. a cell block thing could deduplicate the timestamp/TTL if they are all equal (and delta compress if not).

michoecho replied:
If we are mentioning cell blocks: remember that we don't have to look for one data structure which handles both small and big rows well. We just keep using the radix tree for sufficiently big rows (e.g. more than 128 cells or 10 kiB), but invent a single-buffer layout for smaller rows (which probably account for 99% of the cases where the optimization is actually significant).

mykaul replied:
> Ah, such a pity. a cell block thing could deduplicate the timestamp/TTL if they are all equal (and delta compress if not).

@avikivity - https://github.com/scylladb/scylladb/issues/2954 ?

michoecho replied:
> > Ah, such a pity. a cell block thing could deduplicate the timestamp/TTL if they are all equal (and delta compress if not).
> 
> @avikivity - #2954 ?

Yes, that's the same idea. It's an obvious optimization.

avikivity replied:
I wanted a radix tree of cell blocks.

michoecho replied:
> I wanted a radix tree of cell blocks.

I know, that's why I'm bringing up other ideas before we get too attached to this one.

michoecho commented:
v3:
- Rebased.
- Updated the `managed_bytes` printer in scylla-gdb.py to match the new layout.
- Added `perf_simple_query` and `mutation_footprint_test` results to the cover letter.
- Slightly modified the move constructor to save some instructions.

It should be ready for merge now.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6562/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6562/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7801/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6562/)
#### Build Details:
- Duration: 2 hr 6 min
- Builder: spider6.cloudius-systems.com

bhalevy commented:
@avikivity / @scylladb/scylla-maint please do final review and lets get it merged before it bit-rots

avikivity commented on utils/managed_bytes.hh:
```diff
@@ -157,21 +265,27 @@ public:
+    managed_bytes(managed_bytes&& o) noexcept {
+        // Microoptimization: we use memcpy instead of assigments because
+        // the compiler refuses the merge the load/stores otherwise for some reason.
+        std::memcpy(reinterpret_cast<char*>(this), &o, sizeof(managed_bytes));
```
Don't know if it's completely legal. I'll look the other way.

michoecho replied:
Only trivial types are involved, so copying object representation should be equivalent to default copy constructors.
But I'm not sure if it's completely legal either, union rules are too hard to interpret.

If we are concerned about this, we can just use `char[15]` instead of a union, and access the "union fields" manually via memcpy.
